### DISCUSSION TITLE
Add an ArrayTypecastBehavior in Craft 4

### DISCUSSION THREAD

**Discussion by bencroker** [craftcms/cms#10689] [Ideas] [answered]
Using typed properties in PHP8 is great, until it isn't. I have a plugin that uses an [editableTable](https://github.com/craftcms/cms/blob/4.0/src/templates/_includes/forms/editableTable.twig) form component on the settings page. When populated, it returns an array, but when empty, it returns an empty string.

One option I have is to override the `setAttributes` method in my settings model class.

```php
    // Can be a strict array type
    public array $myArraySetting;

    public function setAttributes($values, $safeOnly = true): void
    {
        // Normalize the array attributes
        foreach ($this->arrayAttributes() as $name) {
            if (isset($values[$name])) {
                $values[$name] = empty($values[$name]) ? [] : (array)$values[$name];
            }
        }

        parent::setAttributes($values, $safeOnly);
    }
```

A nicer option is to register a behavior that typecasts attributes to arrays.

```php
    // Must also allow for a string type, since typecasting happens after setting the value
    public array|string $myArraySetting;

    public function behaviors(): array
    {
        $behaviors = parent::behaviors();
        $behaviors['typecast'] = [
            'class' => ArrayTypecastBehavior::class,
            'attributes' => $this->arrayAttributes(),
        ];

        return $behaviors;
    }
```

I've added the [ArrayTypecastBehavior](https://github.com/putyourlightson/craft-blitz/blob/e834a37dc747f04a2bd1306bfab4243e0b3cd5b9/src/behaviors/ArrayTypecastBehavior.php) to my plugin, but would like to suggest adding it to Craft core.

Taking this a step further, Craft could even do this by default, similar to how it [normalizes date/time attributes](https://github.com/craftcms/cms/blob/a8bace4230763cbe5b3f7dde1df0bf0821861b42/src/base/Model.php#L178-L188).

```php
    public function setAttributes($values, $safeOnly = true): void
    {
        // Normalize the date/time attributes
        foreach ($this->datetimeAttributes() as $name) {
            if (isset($values[$name])) {
                $values[$name] = DateTimeHelper::toDateTime($values[$name]) ?: null;
            }
        }

        // Normalize the array attributes
        foreach ($this->arrayAttributes() as $name) {
            if (isset($values[$name])) {
                $values[$name] = empty($values[$name]) ? [] : (array)$values[$name];
            }
        }

        parent::setAttributes($values, $safeOnly);
    }
```

**Comment by brandonkelly** [comment, marked as answer]
Was just talking to @engram-design about this a couple days ago :)

Just resolved via #10706 – we’re now automatically typecasting all properties with a single, scalar type declaration, from the same places `datetimeAttributes()` is supported. `datetimeAttributes()` is also deprecated now, as it’s redundant for `DateTime` properties now.

**Reply by bencroker** [reply]
Oh that is nice!!

**Reply by engram-design** [reply]
Great work @brandonkelly, beat me to it. Really handy to have a lot of boilerplate casting taken care of. I also think it'll help others transition from Craft 3/4 PHP 7/8.

**Comment by khalwat** [comment]
Dup of my comment on the linked issue from @brandonkelly (because it's been merged already):

Feels like perhaps this should be done via the existing `AttributeTypecastBehavior`? https://www.yiiframework.com/doc/api/2.0/yii-behaviors-attributetypecastbehavior

Or are you eschewing that because it doesn't handle `array` types?

Given that I've already added this behavior to a number of my models, it'd be nice if it were extended or such instead of a new way to do the same (but slightly better) thing?

**Reply by brandonkelly** [reply]
`AttributeTypecastBehavior` is based on the DB schema, whereas this is based on property types directly. You probably can remove most of your existing `AttributeTypecastBehavior` uses at this point, without needing to replace it with anything else assuming you are extending `craft\base\Model`. And if not, it should just require adding this to the constructor:

```php
use craft\helpers\Typecast;

public function __construct(array $config = [])
{
    Typecast::properties(static::class, $config);

    parent::__construct($config);)
}
```

**Reply by khalwat** [reply]
Roger that.

**Comment by khalwat** [comment]
Further thought... the point of typing is to catch errors where you're setting, say, a `string` on an` int` -- all this really does is bring things back to PHP pre-7, where variables are auto-coalesced into the appropriate type. So you'll never catch subtle errors.

I realize it's more convenient but aren't we side-stepping the reason for / benefit of typing to begin with?

I feel like the real issue here is that the value coming to the controller shouldn't be an empty string, it should be an empty array.

**Reply by brandonkelly** [reply]
PHP types are still adding a ton of value – you can be sure that once the value has been set to the property, it’s definitely of the type you want it to be. You don’t have to typecast things manually anymore anywhere, besides when initially setting them.



> I feel like the real issue here is that the value coming to the controller shouldn't be an empty string, it should be an empty array.

It’s not possible to post an empty array, unfortunately, unless you are posting a JSON body via Ajax.

**Comment by khalwat** [comment]
Yeah okay, I think I'll have to walk back my objections now that I understand better what it's doing (had a peek at the code).

It's a better version of `AttributeTypecastBehavior` that leans on typings for the initial model `__construct()` config array. Clever girl.
